x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/TransportPutSamlServiceProviderAction.java

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.idp.action;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.hash.MessageDigests;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.iterable.Iterables;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderDocument;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderIndex;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.stream.Collectors;
+
+public class TransportPutSamlServiceProviderAction
+    extends HandledTransportAction<PutSamlServiceProviderRequest, PutSamlServiceProviderResponse> {
+
+    private final Logger logger = LogManager.getLogger();
+    private final SamlServiceProviderIndex index;
+    private final Clock clock;
+
+    @Inject
+    public TransportPutSamlServiceProviderAction(TransportService transportService, ActionFilters actionFilters,
+                                                 SamlServiceProviderIndex index) {
+        this(transportService, actionFilters, index, Clock.systemUTC());
+    }
+
+    TransportPutSamlServiceProviderAction(TransportService transportService, ActionFilters actionFilters,
+                                                 SamlServiceProviderIndex index, Clock clock) {
+        super(PutSamlServiceProviderAction.NAME, transportService, actionFilters, PutSamlServiceProviderRequest::new);
+        this.index = index;
+        this.clock = clock;
+    }
+
+    @Override
+    protected void doExecute(Task task, final PutSamlServiceProviderRequest request,
+                             final ActionListener<PutSamlServiceProviderResponse> listener) {
+        final SamlServiceProviderDocument document = request.getDocument();
+        if (document.docId != null) {
+            listener.onFailure(new IllegalArgumentException("request document must not have an id [" + document.docId + "]"));
+            return;
+        }
+        index.findByEntityId(document.entityId, ActionListener.wrap(matchingDocuments -> {
+            if (matchingDocuments.isEmpty()) {
+                // derive a document id from the entity id so that don't accidentally create duplicate entities due to a race condition
+                document.docId = deriveDocumentId(document);
+                // force a create in case there are concurrent requests
```

Comment:
Can you elaborate on this comment?

Reply:
If two threads (or 2 nodes) are trying to create the same SP at the same time, then it's possible for them both to write to the index. That is - it's possible for two threads to be in this block simultaneously.

This path assumes that the doc doesn't exist. For now that assumption is weak - all it means is that the creation time is lost. But in the future it might mean more - e.g. If we automatically generate IDP signing credentials on first registration, then have 2 threads thinking they're creating a new SP would do the wrong thing.
If we're in a block that is based on the premise that we are creating the SP for the first time, letting the index enforce that seems helpful.

I could also do the same on the 2nd block - enforce the seqNo/primTerm so that we're only overwriting the doc we expect to. I haven't, but maybe I should?

There's also the theoretical risk of a SHA256 collision. While highly unlikely, it would be an almost impossible issue to diagnose and debug if it did happen. If we do a `INDEX` here, it's possible to overwrite a doc with a matching has (doc-id) but different entity-id. This way it will fail instead, which is not great, but better than silently overwriting data.

Reply:
Makes sense, the 

> This way it will fail instead, which is not great, but better than silently overwriting data.

part is the reasoning I was missing for 

>                  // force a create in case there are concurrent requests 

all clear now

Reply:
I'll update the code comment, to make _the "fail" rather than silently overwrite_ bit explicit.